k8s/pkg/apis/longhorn/v1beta2/objectstore.go

```diff
@@ -0,0 +1,175 @@
+	// and begins resource creation the state is transformed to "starting". The
+	// object store controller observes the resources and once all resources are
+	// up and ready the state transitions to "running".
+	// The state remains in "running" until the object store is deleted, at
+	// which point the controller will clean up the associated resources.
+	//
+	//  │ object store created
+	//  │
+	//  └──►┌───────────┐
+	//      │ unknown   │
+	//  ┌── └───────────┘
+	//  │
+	//  │ controller creates resources
+	//  │
+	//  └──►┌───────────┐
+	//      │ starting  │
+	//  ┌── └───────────┘
+	//  │
+	//  │ controller detects all resources ready
+	//  │
+	//  └──►┌───────────┐
+	//      │ running   │
+	//  ┌── └───────────┘
+	//  │
+	//  │ object endpoint is marked for deletion
+	//  │
+	//  └──►┌───────────┐
+	//      │ stopping  │
+	//  ┌── └───────────┘
+	//  │
+	//  │ controller waits for dependent resources to disappear before removing
+	//  │ the finalizer, thereby letting the object store be deleted
+	//
+	// +optional
+	// +default="Unknown"
+	State ObjectStoreState `json:"state"`
+
+	// A list of addresses where the S3 store is exposed.
+	//
+	// +optional
+	Endpoints []string `json:"endpoints"`
+}
+
+type ObjectStoreCredentials struct {
+	// An access key. This will be the initial access key for the administrative
+	// account.
+	//
+	// +optional
+	AccessKey string `json:"accessKey"`
+
+	// A secret key. This is the secret key corresponding to the access key above.
+	//
+	// +optional
+	SecretKey string `json:"secretKey"`
+}
+
+type ObjectStoreStorageSpec struct {
+	// The initial size of the volume to provision. The volume will be created
+	// such that it can be expanded if resources are available
+	//
+	// +kubebuilder:validation:Required
+	// +kubebuilder:example:=100Gi
+	// +optional
+	Size resource.Quantity `json:"size"`
+
+	// +kubebuilder:validation:Minimum=1
+	// +kubebuilder:default:=3
+	// +optional
+	NumberOfReplicas int `json:"numberOfReplicas"`
+
+	// Replica soft anti affinity of the volume. Set enabled to allow replicas to
+	// be scheduled on the same node
+	//
+	// +optional
+	ReplicaSoftAntiAffinity ReplicaSoftAntiAffinity `json:"replicaSoftAntiAffinity"`
+
+	// Replica zone soft anti affinity of the volume. Set enabled to allow
+	// replicas to be scheduled in the same zone
+	//
+	// +optional
+	ReplicaZoneSoftAntiAffinity ReplicaZoneSoftAntiAffinity `json:"replicaZoneSoftAntiAffinity"`
+
+	// +optional
+	DiskSelector []string `json:"diskSelector"`
+	// +optional
+	NodeSelector []string `json:"nodeSelector"`
+
+	// +optional
+	FromBackup string `json:"fromBackup"`
+
+	// +kubebuilder:validation:Minimum=1
+	// +kubebuilder:default:=2880
+	// +optional
+	StaleReplicaTimeout int `json:"staleReplicaTimeout"`
+}
+
+type ObjectEndpointSpec struct {
+	// +optional
+	Name string `json:"name,omitempty"`
+
+	// +optional
+	DomainName string `json:"domainName"`
+}
+
+type ObjectStoreSpec struct {
```

Comment:
Proposing to have `spec.Image` and `spec.UiImage` fields. Ref https://github.com/longhorn/longhorn-manager/pull/2135/files#diff-2c3ffdbae066b34c3db294fb12d0e69acecd87a77f1100f8789bd885064d46ecR130-R134
 
The reason is that when the user upgrades the Longhorn system to a new version, the newer Longhorn version will most likely come with a new default image for s3gw and s3gw-ui. Longhorn manager cannot immediately swap out the ObjectStore deployment's image because:
1. It will lead to the redeploying of the ObjectEndpoint deployment -> leading to unexpected downtime 
2. Maybe newer s3gw might not compatible with the existing one of the ObjectEndpoint? (I am not very sure about this point though) 

So we might want to keep the existing ObjectStore running with old s3gw image util user upgrade the ObjectStore image. This is similar to how Longhorn volumes' engine image is upgraded after Longhorn system upgrade https://longhorn.io/docs/1.5.1/deploy/upgrade/upgrade-engine/

Reply:
A Longhorn update would not automatically re-create the Deployment. Therefore it would not automatically update the s3gw image in existing ObjectStores anyways.
The only case I can think of where that could happen would be when you do a backup of an ObjectStore on the old Longhorn version, then update the Longhorn installation and then do a restore from that backup.
We could perhaps leverage the ObjectStore's state here to manage controlled upgrades, i.e. the controller would not modify the deployment unless the ObjectStore is in `stopped` state.

Reply:
> A Longhorn update would not automatically re-create the Deployment. Therefore it would not automatically update the s3gw image in existing ObjectStores anyways.

Thank you for pointing out this 

>  i.e. the controller would not modify the deployment unless the ObjectStore is in stopped state.

This can still be problematic. Imagine a user is running a version of s3gw-A that is not compatible to s3gw-B. They stop the ObjectStore and restart it. Suddenly their ObjectStore is broken. 

IMHO, it would be better to let the user to specifically ask for the upgrade by setting the spec.Image fields. When the user performance the update, we have optunity to reject the update if the image are not compatible. 
We can further have a Longhorn setting which automatically upgrade image for ObjectStore to reduce the manual user has to perform. When user set this setting, they agree that they might be some down time during Longhorn upgrade